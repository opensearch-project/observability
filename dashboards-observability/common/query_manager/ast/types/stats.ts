/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * stats chunck types
 */
export interface StatsAggregationChunk {
  function: StatsAggregationFunctionChunk;
  function_alias: string;
}

export interface StatsAggregationFunctionChunk {
  name: string;
  value_expression: string;
  percentile_agg_function: string;
}

export interface GroupField {
  name: string;
}

export interface SpanChunk {
  customLabel: string;
  span_expression: SpanExpressionChunk;
}

export interface SpanExpressionChunk {
  type: string;
  field: string;
  time_unit: string;
  literal_value: string;
}

export interface GroupByChunk {
  group_fields: GroupField[];
  span: SpanChunk | null;
}

export interface statsChunk {
  aggregations: StatsAggregationChunk[];
  groupby: GroupByChunk;
  partitions: ExpressionChunk;
  all_num: ExpressionChunk;
  delim: ExpressionChunk;
  dedup_split_value: ExpressionChunk;
}

export interface ExpressionChunk {
  keyword: string;
  sign: string;
  value: string | number;
}

export interface DataConfigSeries {
  customLabel: string;
  label: string;
  name: string;
  aggregation: string;
}

export interface AggregationConfigurations {
  series: DataConfigSeries[];
  dimensions: GroupField[];
  span: SpanChunk;
  breakdowns: GroupField[];
}

export interface PreviouslyParsedStaleStats {
  partitions: ExpressionChunk;
  all_num: ExpressionChunk;
  delim: ExpressionChunk;
  dedup_split_value: ExpressionChunk;
}
