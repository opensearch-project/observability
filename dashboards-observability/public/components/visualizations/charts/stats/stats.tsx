/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useMemo } from 'react';
import Plotly from 'plotly.js-dist';
import { Plt } from '../../plotly/plot';
import { ThresholdUnitType } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_thresholds';
import { EmptyPlaceholder } from '../../../event_analytics/explorer/visualizations/shared_components/empty_placeholder';
import { last } from 'lodash';
import { AvailabilityUnitType } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_availability';
import { ConfigListEntry } from '../../../../../common/types/explorer';
import {
  STATS_MAX_GRID_LENGTH,
  STATS_GAP_BETWEEN_AXIS,
  DefaultStatsParameters,
  STATS_AXIS_MARGIN,
} from '../../../../../common/constants/explorer';
const {
  DefaultOrientation,
  StatsDefaultTextMode,
  TextSize,
  TextColor,
  TextAlignment,
  ChartType,
} = DefaultStatsParameters;

export const Stats = ({ visualizations, layout, config }: any) => {
  const {
    data,
    metadata: { fields },
  } = visualizations?.data?.rawVizData;

  // data config parametrs
  const { dataConfig = {}, layoutConfig = {} } = visualizations?.data?.userConfigs;
  const dataConfigTab = visualizations.data.rawVizData?.Stats?.dataConfig;
  const dimensions = dataConfigTab?.dimensions
    ? dataConfigTab.dimensions.filter((i: ConfigListEntry) => i.label)
    : [];
  const metrics = dataConfigTab?.metrics
    ? dataConfigTab.metrics.filter((i: ConfigListEntry) => i.label)
    : [];
  const metricsLength = metrics.length;
  const chartType = dataConfig?.chartStyles?.chartType || ChartType;

  if ((chartType === ChartType && dimensions.length === 0) || metricsLength.length === 0)
    return <EmptyPlaceholder icon={visualizations?.vis?.iconType} />;

  // style panel parameters
  const thresholds = dataConfig?.thresholds || [];
  const textSize = dataConfig?.chartStyles?.textSize || TextSize;
  const selectedOrientation = dataConfig?.chartStyles?.orientation || DefaultOrientation;
  const orientation = selectedOrientation === 'auto' || selectedOrientation === 'v' ? 'auto' : 'h';
  const selectedTextMode = dataConfig?.chartStyles?.textMode || StatsDefaultTextMode;
  const textMode =
    selectedTextMode === StatsDefaultTextMode || selectedTextMode === 'values+names'
      ? 'auto'
      : selectedTextMode;
  const textColor = dataConfig?.chartStyles?.textColor?.childColor || TextColor;
  const textAlign = dataConfig?.chartStyles?.textAlign || TextAlignment;

  let autoChartLayout = {
    xaxis: {
      visible: false,
      showgrid: false,
      anchor: 'y1',
      margin: STATS_AXIS_MARGIN,
    },
    yaxis: {
      visible: false,
      showgrid: false,
      anchor: 'x1',
      margin: STATS_AXIS_MARGIN,
    },
  };

  const selectedDimensionsData = dimensions
    .map((dimension: any) => data[dimension.name])
    .reduce((prev, cur) =>
      prev.map((item: string | number, index: number) => `${item},<br>${cur[index]}`)
    );

  const generateLineTraces = () =>
    metrics.map((metric: ConfigListEntry, metricIndex: number) => {
      autoChartLayout = {
        ...autoChartLayout,
        [`yaxis${metricIndex > 0 ? metricIndex + 1 : ''}`]: {
          visible: false,
          showgrid: false,
          anchor: `x${metricIndex > 0 ? metricIndex + 1 : ''}`,
          margin: {
            l: 0,
            r: 0,
            b: 0,
            t: 0,
          },
        },
        [`xaxis${metricIndex > 0 ? metricIndex + 1 : ''}`]: {
          visible: false,
          showgrid: false,
          anchor: `y${metricIndex > 0 ? metricIndex + 1 : ''}`,
          margin: {
            l: 0,
            r: 0,
            b: 0,
            t: 0,
          },
        },
      };

      return {
        x: selectedDimensionsData,
        y: data[metric.label],
        fill: 'tozeroy',
        mode: 'line',
        type: 'scatter',
        name: metric.label,
        ...(metricIndex > 0 && {
          xaxis: `x${metricIndex + 1}`,
          yaxis: `y${metricIndex + 1}`,
        }),
      };
    });

  const [statsData, statsLayout]: Plotly.Data[] = useMemo(() => {
    let calculatedStatsData: Plotly.Data[] = [];
    calculatedStatsData = generateLineTraces();
    // label/text
    let textLabelTrace: any = {
      x: [],
      y: [],
      mode: 'text',
      text: [],
      type: 'scatter',
      xaxis: `x${metricsLength + 1}`,
      yaxis: `y${metricsLength + 1}`,
      textfont: {
        size: textSize,
        color: textColor,
      },
    };
    const repeatedAxis = metricsLength === 1 || orientation === DefaultOrientation ? 'x' : 'y';
    let singleAxis = metricsLength === 1 || orientation === DefaultOrientation ? 'y' : 'x';
    const singleAxisCoords = [
      STATS_MAX_GRID_LENGTH - 1,
      metricsLength === 1 || orientation === DefaultOrientation
        ? STATS_MAX_GRID_LENGTH - 1.4
        : STATS_MAX_GRID_LENGTH - 1,
    ];
    const ZERO_ERROR = metricsLength === 1 || orientation === DefaultOrientation ? 0 : 0.5;
    const isSingleText = textMode === StatsDefaultTextMode ? false : true;
    metrics.forEach((m: ConfigListEntry, index: number) => {
      // for layout of text trace
      if (textMode === StatsDefaultTextMode) {
        textLabelTrace.text.push(`${m.label}`);
        textLabelTrace.text.push(`${data[m.label][data[m.label].length - 1]}`);
      } else if (textMode === 'names') {
        textLabelTrace.text.push(`${m.label}`);
      } else {
        textLabelTrace.text.push(`${data[m.label][data[m.label].length - 1]}`);
      }
      if (index === metricsLength - 1) {
        const textCoords = isSingleText
          ? [STATS_MAX_GRID_LENGTH - 1]
          : [STATS_MAX_GRID_LENGTH - 1, STATS_MAX_GRID_LENGTH - 1 + ZERO_ERROR];
        textLabelTrace[repeatedAxis].push(...textCoords);
      } else {
        if (textLabelTrace[repeatedAxis].length) {
          // covering all cases in between
          const textCoords = isSingleText
            ? [
                textLabelTrace[repeatedAxis][textLabelTrace[repeatedAxis].length - 1] +
                  STATS_MAX_GRID_LENGTH / metricsLength,
              ]
            : [
                textLabelTrace[repeatedAxis][textLabelTrace[repeatedAxis].length - 1] +
                  STATS_MAX_GRID_LENGTH / metricsLength,
                textLabelTrace[repeatedAxis][textLabelTrace[repeatedAxis].length - 1] +
                  STATS_MAX_GRID_LENGTH / metricsLength +
                  ZERO_ERROR,
              ];
          textLabelTrace[repeatedAxis].push(...textCoords);
        } else {
          // for very first metric
          const textCoords = isSingleText
            ? [STATS_MAX_GRID_LENGTH / metricsLength - 1]
            : [
                STATS_MAX_GRID_LENGTH / metricsLength - 1,
                STATS_MAX_GRID_LENGTH / metricsLength - 1 + ZERO_ERROR,
              ];
          textLabelTrace[repeatedAxis].push(...textCoords);
        }
      }

      if (isSingleText) {
        textLabelTrace[singleAxis].push(singleAxisCoords[0]);
      } else {
        textLabelTrace[singleAxis].push(...singleAxisCoords);
      }
    });
    calculatedStatsData = [...calculatedStatsData, textLabelTrace];
    // add layout for text traces
    autoChartLayout = {
      ...autoChartLayout,
      [`xaxis${metricsLength + 1}`]: {
        range: [0, STATS_MAX_GRID_LENGTH],
        showline: false,
        zeroline: false,
        showgrid: false,
      },
      [`yaxis${metricsLength + 1}`]: {
        range: [0, STATS_MAX_GRID_LENGTH],
        showline: false,
        zeroline: false,
        showgrid: false,
      },
    };
    if (thresholds.length) {
      const mapToLine = (list: ThresholdUnitType[] | AvailabilityUnitType[]) =>
        list.map((thr: ThresholdUnitType) =>
          calculatedStatsData
            .filter((i) => i.mode === 'line')
            .map((stat: any, index: number) => ({
              type: 'line',
              x0: data[dimensions[0].label][0],
              y0: thr.value,
              x1: last(data[dimensions[0].label]),
              y1: thr.value,
              xref: `x${index + 1}`,
              yref: `y${index + 1}`,
              name: thr.name || '',
              opacity: 0.7,
              line: {
                color: thr.color,
                width: 3,
                dash: 'dashdot',
              },
            }))
        );

      autoChartLayout = {
        ...autoChartLayout,
        shapes: mapToLine(thresholds).flat(2),
      };
    }
    return [calculatedStatsData, autoChartLayout];
  }, [dimensions, metrics, data, fields, thresholds, orientation, textSize, textMode]);

  const mergedLayout = useMemo(() => {
    return {
      ...layout,
      ...(layoutConfig.layout && layoutConfig.layout),
      showlegend: false,
      margin: STATS_AXIS_MARGIN,
      ...statsLayout,
      grid: {
        ...(orientation === DefaultOrientation
          ? {
              rows: 1,
              columns: metricsLength,
              xgap: STATS_GAP_BETWEEN_AXIS,
            }
          : {
              rows: metricsLength,
              columns: 1,
              ygap: STATS_GAP_BETWEEN_AXIS,
            }),
        pattern: 'independent',
        roworder: 'bottom to top',
      },
      ...(dataConfig?.panelOptions?.title || layoutConfig.layout?.title
        ? {
            title: {
              text: dataConfig?.panelOptions?.title,
              xref: 'paper',
              yref: 'paper',
              y: 1,
            },
          }
        : {}),
    };
  }, [
    data,
    layout,
    layoutConfig.layout,
    dataConfig?.panelOptions?.title,
    orientation,
    metricsLength,
    statsLayout,
    thresholds,
  ]);

  const mergedConfigs = {
    ...config,
    ...(layoutConfig.config && layoutConfig.config),
  };

  return <Plt data={statsData} layout={mergedLayout} config={mergedConfigs} />;
};
