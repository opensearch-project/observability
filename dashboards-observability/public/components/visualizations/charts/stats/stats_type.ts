/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Stats } from './stats';
import { getPlotlySharedConfigs, getPlotlyCategory } from '../shared/shared_configs';
import { LensIconChartLine } from '../../assets/chart_line';
import { PLOTLY_COLOR } from '../../../../../common/constants/shared';
import { VizDataPanel } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/default_vis_editor';
import { ConfigEditor } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/json_editor';
import {
  ConfigThresholds,
  InputFieldItem,
  ButtonGroupItem,
  ConfigChartOptions,
  ColorPalettePicker,
  TextInputFieldItem,
} from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls';
import { ConfigAvailability } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_availability';
import { DefaultStatsParameters } from '../../../../../common/constants/explorer';
import { SINGLE_COLOR_PALETTE } from '../../../../../common/constants/colors';

const sharedConfigs = getPlotlySharedConfigs();
const VIS_CATEGORY = getPlotlyCategory();
const {
  DefaultTextMode,
  DefaultOrientation,
  DefaultChartType,
  DefaultPrecision,
  DefaultTitleSize,
  DefaultValueSize
} = DefaultStatsParameters;

export const createStatsTypeDefinition = (params: any = {}) => ({
  name: 'Stats',
  type: 'stats',
  id: 'stats',
  label: 'Stats',
  fullLabel: 'Stats',
  iconType: 'stats',
  category: VIS_CATEGORY.BASICS,
  selection: {
    dataLoss: 'nothing',
  },
  icon: LensIconChartLine,
  categorYaxis: 'xaxis',
  seriesAxis: 'yaxis',
  chartType: DefaultChartType,
  precisionValue: DefaultPrecision,
  titleSize: DefaultTitleSize,
  valueSize: DefaultValueSize,
  textMode: DefaultTextMode,
  orientation: DefaultOrientation,
  editorConfig: {
    panelTabs: [
      {
        id: 'data-panel',
        name: 'Style',
        mapTo: 'dataConfig',
        editor: VizDataPanel,
        sections: [
          {
            id: 'chart_styles',
            name: 'Chart styles',
            editor: ConfigChartOptions,
            mapTo: 'chartStyles',
            schemas: [
              {
                name: 'Chart type',
                mapTo: 'chartType',
                component: ButtonGroupItem,
                eleType: 'buttons',
                props: {
                  options: [
                    { name: 'Auto', id: DefaultChartType },
                    { name: 'Horizontal', id: 'horizontal' },
                    { name: 'Text mode', id: 'text' },
                  ],
                  defaultSelections: [{ name: 'Auto', id: DefaultChartType }],
                },
              },
              {
                name: 'Orientation',
                component: ButtonGroupItem,
                mapTo: 'orientation',
                eleType: 'buttons',
                props: {
                  options: [
                    { name: 'Auto', id: DefaultOrientation },
                    { name: 'Horizontal', id: 'h' },
                    { name: 'Vertical', id: 'v' },
                  ],
                  defaultSelections: [{ name: 'Auto', id: DefaultOrientation }],
                },
              },
              {
                title: 'Metrics Units',
                name: 'Metric Units',
                component: TextInputFieldItem,
                mapTo: 'metricUnits',
                eleType: 'textInput',
              },
              {
                title: 'Metrics Precision',
                name: 'Metrics Precision',
                component: InputFieldItem,
                mapTo: 'precisionValue',
                eleType: 'input',
              },
              {
                title: 'Title Size',
                name: 'Title Size',
                component: InputFieldItem,
                mapTo: 'titleSize',
                eleType: 'input',
              },
              {
                title: 'Value Size',
                name: 'Value Size',
                component: InputFieldItem,
                mapTo: 'valueSize',
                eleType: 'input',
              },
              {
                name: 'Text Mode',
                component: ButtonGroupItem,
                mapTo: 'textMode',
                eleType: 'buttons',
                props: {
                  options: [
                    { name: 'Auto', id: DefaultTextMode },
                    { name: 'Names', id: 'names' },
                    { name: 'Values', id: 'values' },
                    { name: 'Values + Names', id: 'values+names' },
                  ],
                  defaultSelections: [{ name: 'Values + Names', id: DefaultTextMode }],
                },
              },
              // {
              //   name: 'Text Color',
              //   isSingleSelection: true,
              //   component: ColorPalettePicker,
              //   mapTo: 'textColor',
              //   eleType: 'treemapColorPicker',
              //   options: [
              //     {
              //       value: SINGLE_COLOR_PALETTE,
              //       title: 'Single color',
              //       type: 'text',
              //     },
              //   ],
              //   defaultState: { name: SINGLE_COLOR_PALETTE },
              // },
            ],
          },
          {
            id: 'thresholds',
            name: 'Thresholds',
            editor: ConfigThresholds,
            mapTo: 'thresholds',
            defaultState: [],
            schemas: [],
          },
        ],
      },
      {
        id: 'style-panel',
        name: 'Layout',
        mapTo: 'layoutConfig',
        editor: ConfigEditor,
        content: [],
      },
      {
        id: 'availability-panel',
        name: 'Availability',
        mapTo: 'availabilityConfig',
        editor: ConfigAvailability,
      },
    ],
  },
  visConfig: {
    layout: {
      ...sharedConfigs.layout,
      ...{
        colorway: PLOTLY_COLOR,
        plot_bgcolor: 'rgba(0, 0, 0, 0)',
        paper_bgcolor: 'rgba(0, 0, 0, 0)',
        xaxis: {
          fixedrange: true,
          showgrid: false,
          visible: true,
        },
        yaxis: {
          fixedrange: true,
          showgrid: false,
          visible: true,
        },
      },
    },
    config: {
      ...sharedConfigs.config,
      ...{
        barmode: 'line',
        xaxis: {
          automargin: true,
        },
        yaxis: {
          automargin: true,
        },
      },
    },
  },
  component: Stats,
});
