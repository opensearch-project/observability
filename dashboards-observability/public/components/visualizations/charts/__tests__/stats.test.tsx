/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { Stats } from '../stats/stats';
import {
  LAYOUT_CONFIG,
  STATS_TEST_VISUALIZATIONS_DATA,
} from '../../../../../test/event_analytics_constants';

describe('Stats component', () => {
  configure({ adapter: new Adapter() });

  it('Renders stats component', async () => {
    const wrapper = mount(
      <Stats visualizations={STATS_TEST_VISUALIZATIONS_DATA} layout={LAYOUT_CONFIG} config={true} />
    );

    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
