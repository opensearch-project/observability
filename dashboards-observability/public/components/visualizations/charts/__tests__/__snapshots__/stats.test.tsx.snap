// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats component Renders stats component 1`] = `
<Stats
  config={true}
  layout={
    Object {
      "colorway": Array [
        "#8C55A3",
      ],
      "height": 220,
      "margin": Object {
        "b": 15,
        "l": 60,
        "pad": 0,
        "r": 10,
        "t": 30,
      },
      "showlegend": true,
    }
  }
  visualizations={
    Object {
      "data": Object {
        "appData": Object {
          "fromApp": false,
        },
        "defaultAxes": Object {},
        "indexFields": Object {
          "availableFields": Array [
            Object {
              "name": "agent",
              "type": "string",
            },
            Object {
              "name": "bytes",
              "type": "long",
            },
            Object {
              "name": "clientip",
              "type": "ip",
            },
            Object {
              "name": "event",
              "type": "struct",
            },
            Object {
              "name": "extension",
              "type": "string",
            },
            Object {
              "name": "geo",
              "type": "struct",
            },
            Object {
              "name": "host",
              "type": "string",
            },
            Object {
              "name": "index",
              "type": "string",
            },
            Object {
              "name": "ip",
              "type": "ip",
            },
            Object {
              "name": "machine",
              "type": "struct",
            },
            Object {
              "name": "memory",
              "type": "double",
            },
            Object {
              "name": "message",
              "type": "string",
            },
            Object {
              "name": "phpmemory",
              "type": "long",
            },
            Object {
              "name": "referer",
              "type": "string",
            },
            Object {
              "name": "request",
              "type": "string",
            },
            Object {
              "name": "response",
              "type": "string",
            },
            Object {
              "name": "tags",
              "type": "string",
            },
            Object {
              "name": "timestamp",
              "type": "timestamp",
            },
            Object {
              "name": "url",
              "type": "string",
            },
            Object {
              "name": "utc_time",
              "type": "timestamp",
            },
          ],
          "queriedFields": Array [
            Object {
              "name": "double_per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "host",
              "type": "text",
            },
            Object {
              "name": "ip_count",
              "type": "integer",
            },
            Object {
              "name": "per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "resp_code",
              "type": "text",
            },
            Object {
              "name": "sum_bytes",
              "type": "long",
            },
          ],
          "selectedFields": Array [],
          "unselectedFields": Array [],
        },
        "query": Object {},
        "rawVizData": Object {
          "data": Object {
            "count()": Array [
              154,
              1753,
              116,
              468,
              1964,
              219,
            ],
            "tags": Array [
              "error",
              "info",
              "login",
              "security",
              "success",
              "warning",
            ],
          },
          "jsonData": Array [
            Object {
              "count()": 154,
              "tags": "error",
            },
            Object {
              "count()": 1753,
              "tags": "info",
            },
            Object {
              "count()": 116,
              "tags": "login",
            },
            Object {
              "count()": 468,
              "tags": "security",
            },
            Object {
              "count()": 1964,
              "tags": "success",
            },
            Object {
              "count()": 219,
              "tags": "warning",
            },
          ],
          "metadata": Object {
            "fields": Array [
              Object {
                "name": "count()",
                "type": "integer",
              },
              Object {
                "name": "tags",
                "type": "text",
              },
            ],
            "size": 6,
            "status": 200,
          },
        },
        "userConfigs": Object {
          "dataConfig": Object {
            "valueOptions": Object {
              "dimensions": Array [
                Object {
                  "label": "tags",
                  "name": "tags",
                  "type": "text",
                },
              ],
              "metrics": Array [
                Object {
                  "label": "count()",
                  "name": "count()",
                  "side": "left",
                  "type": "integer",
                },
              ],
            },
          },
        },
      },
      "vis": Object {
        "category": "Visualizations",
        "categoryaxis": "xaxis",
        "charttype": "auto",
        "component": [Function],
        "editorconfig": Object {
          "panelTabs": Array [
            Object {
              "editor": [Function],
              "id": "data-panel",
              "mapTo": "dataConfig",
              "name": "Style",
              "sections": Array [
                Object {
                  "editor": [Function],
                  "id": "tooltip_options",
                  "mapTo": "tooltipOptions",
                  "name": "Tooltip options",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "mapTo": "tooltipMode",
                      "name": "Tooltip mode",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "show",
                            "name": "Show",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "show",
                            "name": "Show",
                          },
                          Object {
                            "id": "hidden",
                            "name": "Hidden",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": null,
                      "mapTo": "tooltipText",
                      "name": "Tooltip text",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "all",
                            "name": "All",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "all",
                            "name": "All",
                          },
                          Object {
                            "id": "x",
                            "name": "Dimension",
                          },
                          Object {
                            "id": "y",
                            "name": "Metrics",
                          },
                        ],
                      },
                    },
                  ],
                },
                Object {
                  "editor": [Function],
                  "id": "chart_styles",
                  "mapTo": "chartStyles",
                  "name": "Chart styles",
                  "schemas": Array [
                    Object {
                      "component": [Function],
                      "eleType": "buttons",
                      "mapTo": "chartType",
                      "name": "Chart type",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "auto",
                            "name": "Auto",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "auto",
                            "name": "Auto",
                          },
                          Object {
                            "id": "horizontal",
                            "name": "Horizontal",
                          },
                          Object {
                            "id": "text",
                            "name": "Text mode",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": [Function],
                      "eleType": "buttons",
                      "mapTo": "orientation",
                      "name": "Orientation",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "auto",
                            "name": "Auto",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "auto",
                            "name": "Auto",
                          },
                          Object {
                            "id": "h",
                            "name": "Horizontal",
                          },
                          Object {
                            "id": "v",
                            "name": "Vertical",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": [Function],
                      "eleType": "textInput",
                      "mapTo": "metricUnits",
                      "name": "Metric units",
                      "title": "Metrics units",
                    },
                    Object {
                      "component": [Function],
                      "eleType": "input",
                      "mapTo": "precisionValue",
                      "name": "Metrics precision",
                      "props": Object {
                        "minLimit": 0,
                      },
                      "title": "Metrics precision",
                    },
                    Object {
                      "component": [Function],
                      "eleType": "input",
                      "mapTo": "titleSize",
                      "name": "Title size",
                      "title": "Title size",
                    },
                    Object {
                      "component": [Function],
                      "eleType": "input",
                      "mapTo": "valueSize",
                      "name": "Value size",
                      "title": "Value size",
                    },
                    Object {
                      "component": [Function],
                      "eleType": "buttons",
                      "mapTo": "textMode",
                      "name": "Text mode",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "id": "auto",
                            "name": "Values + Names",
                          },
                        ],
                        "options": Array [
                          Object {
                            "id": "auto",
                            "name": "Auto",
                          },
                          Object {
                            "id": "names",
                            "name": "Names",
                          },
                          Object {
                            "id": "values",
                            "name": "Values",
                          },
                          Object {
                            "id": "values+names",
                            "name": "Values + Names",
                          },
                        ],
                      },
                    },
                  ],
                },
                Object {
                  "defaultState": Array [
                    Object {
                      "color": "#3CA1C7",
                      "isReadOnly": true,
                      "name": "Base",
                      "thid": "random_html_id",
                      "value": 0,
                    },
                  ],
                  "editor": [Function],
                  "id": "thresholds",
                  "mapTo": "thresholds",
                  "name": "Thresholds",
                  "schemas": Array [],
                },
              ],
            },
            Object {
              "content": Array [],
              "editor": [Function],
              "id": "style-panel",
              "mapTo": "layoutConfig",
              "name": "Layout",
            },
            Object {
              "editor": [Function],
              "id": "availability-panel",
              "mapTo": "availabilityConfig",
              "name": "Availability",
            },
          ],
        },
        "fulllabel": "Stats",
        "icon": [Function],
        "icontype": "stats",
        "id": "stats",
        "label": "Stats",
        "name": "Stats",
        "orientation": "auto",
        "precisionvalue": 1,
        "selection": Object {
          "dataLoss": "nothing",
        },
        "seriesaxis": "yaxis",
        "textmode": "auto",
        "titlesize": 30,
        "type": "stats",
        "valuesize": 80,
        "visconfig": Object {
          "config": Object {
            "displaylogo": false,
            "responsive": true,
          },
          "layout": Object {
            "colorway": Array [
              "#3CA1C7",
              "#8C55A3",
              "#DB748A",
              "#F2BE4B",
              "#68CCC2",
              "#2A7866",
              "#843769",
              "#374FB8",
              "#BD6F26",
              "#4C636F",
            ],
            "height": 1180,
            "legend": Object {
              "orientation": "v",
              "traceorder": "normal",
            },
            "margin": Object {
              "b": 30,
              "l": 60,
              "pad": 0,
              "r": 30,
              "t": 50,
            },
            "paper_bgcolor": "rgba(0, 0, 0, 0)",
            "plot_bgcolor": "rgba(0, 0, 0, 0)",
            "showlegend": true,
            "xaxis": Object {
              "fixedrange": true,
              "showgrid": false,
              "visible": true,
            },
            "yaxis": Object {
              "fixedrange": true,
              "showgrid": false,
              "visible": true,
            },
          },
        },
      },
    }
  }
>
  <Plt
    config={Object {}}
    data={
      Array [
        Object {
          "fill": "tozeroy",
          "fillcolor": "rgba(60,161,199,0.2)",
          "hoverinfo": "all",
          "line": Object {
            "color": "#3CA1C7",
          },
          "metricValue": "219.0",
          "mode": "lines",
          "name": "count()",
          "type": "scatter",
          "x": Array [
            "error",
            "info",
            "login",
            "security",
            "success",
            "warning",
          ],
          "y": Array [
            154,
            1753,
            116,
            468,
            1964,
            219,
          ],
        },
      ]
    }
    layout={
      Object {
        "annotations": Array [
          Object {
            "font": Object {
              "color": "rgb(0,0,0)",
              "family": "Roboto",
              "size": 28.5,
            },
            "metricValue": "219.0",
            "showarrow": false,
            "text": "count()",
            "type": "name",
            "x": 0,
            "xanchor": "left",
            "xref": "paper",
            "y": 1,
            "yanchor": "bottom",
            "yref": "paper",
          },
          Object {
            "font": Object {
              "color": "#3CA1C7",
              "family": "Roboto",
              "size": 73.6,
            },
            "metricValue": "219.0",
            "showarrow": false,
            "text": "<b>219.0</b>",
            "type": "value",
            "x": 0,
            "xanchor": "left",
            "xref": "paper",
            "y": 1,
            "yanchor": "top",
            "yref": "paper",
          },
        ],
        "colorway": Array [
          "#8C55A3",
        ],
        "grid": Object {
          "columns": 1,
          "pattern": "independent",
          "roworder": "bottom to top",
          "rows": 1,
          "xgap": 0.01,
        },
        "height": 220,
        "margin": Object {
          "b": 0,
          "l": 0,
          "r": 0,
          "t": 80,
        },
        "showlegend": false,
        "title": "",
        "xaxis": Object {
          "anchor": "y",
          "layoutFor": "count()",
          "showgrid": false,
          "visible": false,
        },
        "yaxis": Object {
          "anchor": "x",
          "layoutFor": "count()",
          "range": Array [
            0,
            2946,
          ],
          "showgrid": false,
          "visible": false,
        },
      }
    }
  >
    <PlotlyComponent
      config={
        Object {
          "displayModeBar": false,
        }
      }
      data={
        Array [
          Object {
            "fill": "tozeroy",
            "fillcolor": "rgba(60,161,199,0.2)",
            "hoverinfo": "all",
            "line": Object {
              "color": "#3CA1C7",
            },
            "metricValue": "219.0",
            "mode": "lines",
            "name": "count()",
            "type": "scatter",
            "x": Array [
              "error",
              "info",
              "login",
              "security",
              "success",
              "warning",
            ],
            "y": Array [
              154,
              1753,
              116,
              468,
              1964,
              219,
            ],
          },
        ]
      }
      debug={false}
      divId="explorerPlotComponent"
      layout={
        Object {
          "annotations": Array [
            Object {
              "font": Object {
                "color": "rgb(0,0,0)",
                "family": "Roboto",
                "size": 28.5,
              },
              "metricValue": "219.0",
              "showarrow": false,
              "text": "count()",
              "type": "name",
              "x": 0,
              "xanchor": "left",
              "xref": "paper",
              "y": 1,
              "yanchor": "bottom",
              "yref": "paper",
            },
            Object {
              "font": Object {
                "color": "#3CA1C7",
                "family": "Roboto",
                "size": 73.6,
              },
              "metricValue": "219.0",
              "showarrow": false,
              "text": "<b>219.0</b>",
              "type": "value",
              "x": 0,
              "xanchor": "left",
              "xref": "paper",
              "y": 1,
              "yanchor": "top",
              "yref": "paper",
            },
          ],
          "autosize": true,
          "barmode": "stack",
          "colorway": Array [
            "#8C55A3",
          ],
          "grid": Object {
            "columns": 1,
            "pattern": "independent",
            "roworder": "bottom to top",
            "rows": 1,
            "xgap": 0.01,
          },
          "height": 220,
          "hovermode": "closest",
          "legend": Object {
            "orientation": "h",
            "traceorder": "normal",
          },
          "margin": Object {
            "b": 0,
            "l": 0,
            "r": 0,
            "t": 80,
          },
          "showlegend": false,
          "title": "",
          "xaxis": Object {
            "anchor": "y",
            "layoutFor": "count()",
            "showgrid": false,
            "visible": false,
          },
          "yaxis": Object {
            "anchor": "x",
            "layoutFor": "count()",
            "range": Array [
              0,
              2946,
            ],
            "showgrid": false,
            "visible": false,
          },
        }
      }
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
      useResizeHandler={true}
    >
      <div
        id="explorerPlotComponent"
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </PlotlyComponent>
  </Plt>
</Stats>
`;
